crux_core/src/bridge/serde.rs

```diff
@@ -0,0 +1,49 @@
+use std::error::Error;
+
+pub use serde::{Deserialize, Serialize};
+
+/// A basic serializer for data across the FFI boundary. This allows
+/// you to choose a different serialization format than bincode.
+pub trait Serializer: Clone {
+    type Error: Error;
+
+    fn serialize<T>(&self, data: T) -> Result<Vec<u8>, Self::Error>
+    where
+        T: Serialize;
+}
+
+/// A basic derserializer for data across the FFI boundary. This allows
+/// you to choose a different serialization format than bincode.
+pub trait Deserializer {
+    type Error: Error;
+
+    fn deserialize<'de, T>(&self, data: &'de [u8]) -> Result<T, Self::Error>
+    where
+        T: Deserialize<'de>;
+}
```

Comment:
These feel like I'm reinventing the wheel somehow. Any better suggestions?

Reply:
Why not just use serde?

Reply:
I couldn't quite find the right thing in there when I looked. Their `Serializer` and `Deserializer` are quite complicated and I'm not entirely sure they do the whole thing we need, i.e. "A thing that turns types into `Vec<u8>`", theirs goes into the data format, if I understand right. 

It's highly likely I'm missing something though, it's been a long day. If you could point me at the right trait from their end, I would love not to define our own.

Reply:
You'd probably need to update `BridgeWithSerializer::process_event`/`handle_response` to take a `Deserializer`  & `Serializer` pair instead of taking a byte buffer and outputting a byte buffer.

Then it's just `data.deserialize(deserializer)` for the input and `data.serialize(serializer)` on the other end.

The plain bridge would just wrap its byte buffer input in one of these: https://docs.rs/bincode/1.3.3/bincode/de/struct.Deserializer.html and pass in one of these: https://docs.rs/bincode/1.3.3/bincode/struct.Serializer.html wrapping another buffer for the return value.

Not sure how much better it'd be, the current approach isn't the worst

Reply:
We tried this for a while, but the problem is the resolve callbacks now become generic over the type of deserializer, which bubbles all the way up to the bridge. That's fine for the generic bridge, but for the current, bridge, we need to be able to express that concrete type of deserializer, and unfortunately, in bincode's case, the type is dependent on the data source down to its lifetime. The type is:

```rust
pub struct Deserializer<R, O: Options> {
    pub(crate) reader: R,
    options: O,
}
```

where `R`, in our case is a `SliceReader` and that's defined as:

```rust
pub struct SliceReader<'storage> {
    slice: &'storage [u8],
}
```

Reply:
Don't think this mild improvement is worth changing the `Bridge` type (and I'm not 100% sure what that would even look like)